osu.Game/Database/RealmContextFactory.cs

```diff
@@ -0,0 +1,160 @@
+        /// </summary>
+        private readonly object writeLock = new object();
+
+        private ThreadLocal<bool> refreshCompleted = new ThreadLocal<bool>();
+
+        private bool rollbackRequired;
+
+        private int currentWriteUsages;
+
+        private Transaction currentWriteTransaction;
+
+        public RealmContextFactory(Storage storage)
+        {
+            this.storage = storage;
+
+            threadContexts = new ThreadLocal<Realm>(createContext, true);
+
+            // creating a context will ensure our schema is up-to-date and migrated.
+            var realm = Get();
+            Logger.Log($"Opened realm \"{realm.Config.DatabasePath}\" at version {realm.Config.SchemaVersion}");
+        }
+
+        private void onMigration(Migration migration, ulong lastSchemaVersion)
+        {
+        }
+
+        private static readonly GlobalStatistic<int> reads = GlobalStatistics.Get<int>("Realm", "Get (Read)");
+        private static readonly GlobalStatistic<int> writes = GlobalStatistics.Get<int>("Realm", "Get (Write)");
+        private static readonly GlobalStatistic<int> commits = GlobalStatistics.Get<int>("Realm", "Commits");
+        private static readonly GlobalStatistic<int> rollbacks = GlobalStatistics.Get<int>("Realm", "Rollbacks");
+        private static readonly GlobalStatistic<int> contexts_open = GlobalStatistics.Get<int>("Realm", "Contexts (Open)");
+        private static readonly GlobalStatistic<int> contexts_created = GlobalStatistics.Get<int>("Realm", "Contexts (Created)");
+
+        /// <summary>
+        /// Get a context for the current thread for read-only usage.
+        /// If a <see cref="RealmWriteUsage"/> is in progress, the existing write-safe context will be returned.
+        /// </summary>
+        public Realm Get()
+        {
+            reads.Value++;
+            return getContextForCurrentThread();
+        }
+
+        /// <summary>
+        /// Request a context for write usage. Can be consumed in a nested fashion (and will return the same underlying context).
+        /// This method may block if a write is already active on a different thread.
+        /// </summary>
+        /// <returns>A usage containing a usable context.</returns>
+        public RealmWriteUsage GetForWrite()
+        {
+            writes.Value++;
+            Monitor.Enter(writeLock);
+            Realm context;
+
+            try
+            {
+                context = getContextForCurrentThread();
+
+                currentWriteTransaction ??= context.BeginWrite();
+            }
+            catch
+            {
+                // retrieval of a context could trigger a fatal error.
+                Monitor.Exit(writeLock);
+                throw;
+            }
+
+            Interlocked.Increment(ref currentWriteUsages);
+
+            return new RealmWriteUsage(context, usageCompleted) { IsTransactionLeader = currentWriteTransaction != null && currentWriteUsages == 1 };
```

Comment:
seems slightly dangerous (what if one of these branches randomly breaks and then there's suddenly no leader?) - would probably just check the usage counter and make the `currentWriteTransaction != null` part a super fatal assertion